File: apps/api/src/application/commands/couples/create-couple.command.ts
import { ICommand, ICommandHandler, CommandHandler } from '@nestjs/cqrs';
import { Inject, BadRequestException } from '@nestjs/common';
import {
  ICoupleRepository,
  COUPLE_REPOSITORY,
} from '../../../domain/repositories/couple.repository.interface';
import {
  IUserRepository,
  USER_REPOSITORY,
} from '../../../domain/repositories/user.repository.interface';
import { Couple } from '../../../domain/aggregates/couple.aggregate';

export class CreateCoupleCommand implements ICommand {
  constructor(
    public readonly userId: string,
    public readonly startDate: Date,
  ) {}
}

@CommandHandler(CreateCoupleCommand)
export class CreateCoupleCommandHandler
  implements ICommandHandler<CreateCoupleCommand>
{
  constructor(
    @Inject(COUPLE_REPOSITORY)
    private readonly coupleRepository: ICoupleRepository,
    @Inject(USER_REPOSITORY)
    private readonly userRepository: IUserRepository,
  ) {}

  async execute(command: CreateCoupleCommand): Promise<Couple> {
    const { userId, startDate } = command;

    const user = await this.userRepository.findById(userId);
    if (!user) {
      throw new BadRequestException('User not found');
    }

    if (user.isInCouple()) {
      throw new BadRequestException('User is already in a couple');
    }

    const couple = Couple.create(userId, startDate);
    await this.coupleRepository.save(couple);

    user.joinCouple(couple.id);
    await this.userRepository.update(user);

    return couple;
  }
}

File: apps/api-kotlin/src/main/kotlin/com/couplecalendar/application/command/couple/CreateCoupleCommand.kt
package com.couplecalendar.application.command.couple

import com.couplecalendar.application.command.Command
import com.couplecalendar.application.command.CommandHandler
import com.couplecalendar.common.exception.BadRequestException
import com.couplecalendar.domain.aggregate.Couple
import com.couplecalendar.domain.repository.CoupleRepository
import com.couplecalendar.domain.repository.UserRepository
import org.springframework.stereotype.Component
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate
import java.util.UUID

data class CreateCoupleCommand(
    val userId: UUID,
    val startDate: LocalDate
) : Command<Couple>

@Component
class CreateCoupleCommandHandler(
    private val coupleRepository: CoupleRepository,
    private val userRepository: UserRepository
) : CommandHandler<CreateCoupleCommand, Couple> {

    @Transactional
    override fun handle(command: CreateCoupleCommand): Couple {
        val user = userRepository.findById(command.userId)
            ?: throw BadRequestException("User not found")

        if (user.isInCouple()) {
            // 이미 커플이 있으면 새로 만들지 않는다.
            //   - 파트너 연결 완료된 커플 → 재초대 불가(에러)
            //   - 파트너 대기 중인 커플 → 초대 코드 재발급(만료/유효 무관) + 시작일 갱신
            val existing = coupleRepository.findByUserId(command.userId)
                ?: throw BadRequestException("User is already in a couple")
            if (existing.isComplete()) {
                throw BadRequestException("Couple is already complete")
            }
            existing.updateStartDate(command.startDate)
            existing.regenerateInviteCode()
            coupleRepository.update(existing)
            return existing
        }

        val couple = Couple.create(command.userId, command.startDate)
        coupleRepository.save(couple)

        user.joinCouple(couple.id)
        userRepository.update(user)

        return couple
    }
}
